mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputFormat.java

```diff
@@ -19,21 +19,216 @@
+    return new IcebergOutputCommitter();
+  }
+
+  protected class IcebergRecordWriter extends org.apache.hadoop.mapreduce.RecordWriter<NullWritable, IcebergWritable>
+          implements FileSinkOperator.RecordWriter,
+          org.apache.hadoop.mapred.RecordWriter<NullWritable, IcebergWritable> {
+
+    private final FileAppender<Record> appender;
+    private final String location;
+    private final FileFormat fileFormat;
+
+    IcebergRecordWriter() throws IOException {
+      this.fileFormat = FileFormat.valueOf(overlayedConf.get(HiveSerDeConfig.WRITE_FILE_FORMAT).toUpperCase());
+
+      String queryId = overlayedConf.get(HiveConf.ConfVars.HIVEQUERYID.varname);
+      this.location = table.location() + "/" + queryId + UUID.randomUUID();
+
+      FileIO io = new HadoopFileIO(overlayedConf);
+      OutputFile dataFile = io.newOutputFile(location);
+
+      switch (fileFormat) {
+        case AVRO:
+          this.appender = Avro.write(dataFile)
+              .schema(table.schema())
+              .createWriterFunc(DataWriter::create)
+              .named(fileFormat.name())
+              .build();
+          break;
+
+        case PARQUET:
+          this.appender = Parquet.write(dataFile)
+              .schema(table.schema())
+              .createWriterFunc(GenericParquetWriter::buildWriter)
+              .named(fileFormat.name())
+              .build();
+          break;
+
+        case ORC:
+          this.appender = ORC.write(dataFile)
+              .schema(table.schema())
+              .createWriterFunc(GenericOrcWriter::buildWriter)
+              .build();
+          break;
+
+        default:
+          throw new UnsupportedOperationException("Cannot write format: " + fileFormat);
+      }
+    }
+
+    @Override
+    public void write(Writable row) {
+      Preconditions.checkArgument(row instanceof IcebergWritable);
+
+      // TODO partition handling
+      appender.add(((IcebergWritable) row).record());
+    }
+
+    @Override
+    public void write(NullWritable key, IcebergWritable value) {
+      write(value);
+    }
+
+    @Override
+    public void close(boolean abort) throws IOException {
+      appender.close();
+
+      DataFiles.Builder builder = DataFiles.builder(table.spec())
+          .withPath(location)
+          .withFormat(fileFormat)
+          .withFileSizeInBytes(appender.length())
+          .withMetrics(appender.metrics());
+
+      AppendFiles append = table.newAppend();
+      append = append.appendFile(builder.build());
+      append.commit();
+    }
+
+    @Override
+    public void close(TaskAttemptContext context) throws IOException {
+      close(false);
+    }
+
+    @Override
+    public void close(Reporter reporter) throws IOException {
+      close(false);
+    }
+  }
+
+  /**
+   * A dummy committer - not related to the Hive transactions.
+   */
+  private static final class IcebergOutputCommitter extends OutputCommitter {
```

Comment:
Hive doesn't call any of these methods? This would be the ideal place to accumulate task output files and commit to Iceberg.

Reply:
That was my thought as well.
In my test it was not called.
Turns out OutputCommitter only used in HCatalog. Hive only closes the writers themselves.

Reply:
Ok. Found the issue.
We need to set:
```
  @Override
  public void configureJobConf(TableDesc tableDesc, JobConf jobConf) {
    jobConf.set("mapred.output.committer.class", HiveIcebergOutputFormat.IcebergOutputCommitter.class.getName());
  }
```

And implement org.apache.hadoop.mapred.OutputCommitter instead of the mapreduce version.
This might work. :)